File: Android/app/src/main/java/com/dkaluta/prosary/models/Prayer.kt
package com.dkaluta.prosary.models

import android.content.Context
import java.util.UUID

/** A saved, user-configurable prayer session. [kind] selects the Rosary, the Jesus Prayer, or
 * the generic bundle path; adding an ordinary devotion means adding a bundle, not a new case. */
data class Prayer(
    val id: String = UUID.randomUUID().toString(),
    var name: String = "My Prayer",
    var kind: PrayerKind = PrayerKind.Rosary,

    /** The primary configuration for its devotion. At most one per
     * ([kind], [customDevotionId]) at a time. */
    var isDefault: Boolean = false,

    /** Prayer language for this configuration. [LanguageCatalog.defaultSentinel] means follow the
     * app-level default language setting. */
    var languageCode: String = LanguageCatalog.defaultSentinel,

    // Kind-specific options — populate the relevant class when creating a Prayer.
    var rosary: RosaryOptions = RosaryOptions(),
    var jesusPrayer: JesusPrayerOptions = JesusPrayerOptions(),
    /** The bundle id (e.g. "trisagion") whose `devotion.json` defines this configuration's step sequence
     * — populated only when [kind] is [PrayerKind.Custom], null otherwise. See
     * [com.dkaluta.prosary.engine.PrayerEngine.buildCustomDevotionSteps]. */
    var customDevotionId: String? = null,
    /** Which of the bundle's variants (alternate step-sets, e.g. the Stations' traditional vs.
     * scriptural forms) this configuration prays. Null = the bundle's default variant; only
     * meaningful when kind == Custom and the bundle declares variants. */
    var variantId: String? = null,
    /** Multi-day ("days"-type) devotions: the day this configuration opens on, 0-based; advances
     * when a day's session finishes (clamped by the engine). Null = day 1. */
    var dayIndex: Int? = null,

    /** This configuration's choices for the bundle's `options.json` options, keyed by option key —
     * "true"/"false" for toggles, a case id for choices. Only overrides: an absent key means the
     * option's declared default. Only meaningful when [kind] == [PrayerKind.Custom]. */
    var customOptions: Map<String, String> = emptyMap(),

    /** Daily reminders for this configuration. Scheduled via
     * [com.dkaluta.prosary.reminders.ReminderScheduler]. */
    var reminders: List<PrayerReminder> = emptyList(),
) {
    val isNotDefault: Boolean get() = !isDefault
    val resolvedLanguageCode: String get() = LanguageCatalog.resolve(languageCode).code
    val languageNativeName: String get() = LanguageCatalog.resolve(languageCode).nativeName

    /** Display string for list rows — shows "Default (Latina)" for the sentinel, plain name otherwise. */
    fun languageDisplayName(context: Context): String =
        if (languageCode == LanguageCatalog.defaultSentinel) {
            context.getString(
                com.dkaluta.prosary.R.string.language_default_parenthesized,
                LanguageCatalog.resolve(languageCode).nativeName,
            )
        } else {
            LanguageCatalog.resolve(languageCode).nativeName
        }
}

File: iOS/Prosary/Models/Prayer.swift

//
//  Prayer.swift
//  Prosary
//
//  A saved, user-configurable prayer session. `kind` selects the Rosary, the Jesus Prayer, or
//  the generic bundle path; adding an ordinary devotion means adding a bundle, not a new case.
//

import Foundation

struct Prayer: Identifiable, Hashable, Codable {
  var id = UUID()
  var name: String = "My Prayer"
  var kind: PrayerKind = .rosary

  /// The primary configuration for its devotion. At most one per
  /// (`kind`, `customDevotionId`) at a time.
  var isDefault: Bool = false

  /// Prayer language for this configuration. `LanguageCatalog.defaultSentinel` means follow the
  /// app-level default language setting.
  var languageCode: String = LanguageCatalog.defaultSentinel

  // Kind-specific options — populate the relevant struct when creating a Prayer.
  var rosary: RosaryOptions = .init()
  var jesusPrayer: JesusPrayerOptions = .init()

  /// The bundle id (e.g. `"angelus"`) whose `devotion.json` defines this configuration's step
  /// sequence — populated only when `kind == .custom`, nil otherwise. See `PrayerEngine.
  /// buildCustomDevotionSteps`.
  var customDevotionId: String? = nil

  /// Which of the bundle's variants (alternate step-sets, e.g. the Stations' traditional vs.
  /// scriptural forms) this configuration prays. Nil = the bundle's default variant; only
  /// meaningful when `kind == .custom` and the bundle declares variants.
  var variantId: String? = nil

  /// Multi-day ("days"-type) devotions: the day this configuration opens on, 0-based. Advances
  /// when a day's session finishes; clamped by the engine so a completed novena re-prays its
  /// last day. Nil (the CloudKit-safe optional default) means day 1. Only meaningful when
  /// `kind == .custom` and the bundle's devotion is days-type.
  var dayIndex: Int? = nil

  /// This configuration's choices for the bundle's `options.json` options, keyed by option key —
  /// "true"/"false" for toggles, a case id for choices. Only overrides: an absent key means the
  /// option's declared default. Only meaningful when `kind == .custom`.
  var customOptions: [String: String] = [:]

  /// Daily reminders for this configuration. Scheduled via `ReminderScheduler`.
  var reminders: [PrayerReminder] = []

  var isNotDefault: Bool { !isDefault }
  var resolvedLanguageCode: String { LanguageCatalog.resolve(languageCode).code }
  var languageNativeName: String { LanguageCatalog.resolve(languageCode).nativeName }

  /// Display string for list rows — shows "Default (Latina)" for sentinel, plain name otherwise.
  var languageDisplayName: String {
    languageCode == LanguageCatalog.defaultSentinel
      ? "Default (\(LanguageCatalog.resolve(languageCode).nativeName))"
      : LanguageCatalog.resolve(languageCode).nativeName
  }
}
